### PR TITLE
OSDOCS#9001: Adding vSphere credentials to Agent

### PR DESCRIPTION
**Pull request by skopacz1**
[OSDOCS-9001](https://issues.redhat.com/browse/OSDOCS-9001)

Versions: 4.15+

This PR adds existing vSphere credential parameters for the install-config that are now available for the Agent-based Installer

QE review:
- [x] QE has approved this change.

Preview: 

- [vSphere parameters on the Agent page](https://70551--ocpdocs-pr.netlify.app/openshift-enterprise/latest/installing/installing_with_agent_based_installer/installation-config-parameters-agent#installation-configuration-parameters-additional-vsphere_installation-config-parameters-agent)
- [Creating the preferred configuration inputs](https://70551--ocpdocs-pr.netlify.app/openshift-enterprise/latest/installing/installing_with_agent_based_installer/installing-with-agent-based-installer#installing-ocp-agent-inputs_installing-with-agent-based-installer) (Additional resources section at end of procedure)
- [vSphere parameters on the vSphere page](https://70551--ocpdocs-pr.netlify.app/openshift-enterprise/latest/installing/installing_vsphere/installation-config-parameters-vsphere#installation-configuration-parameters-additional-vsphere_installation-config-parameters-vsphere) (should be unchanged)

Existing docs:

- [Agent config parameters](https://docs.openshift.com/container-platform/4.14/installing/installing_with_agent_based_installer/installation-config-parameters-agent.html)
- [vSphere config parameters](https://docs.openshift.com/container-platform/4.14/installing/installing_vsphere/installation-config-parameters-vsphere.html#installation-configuration-parameters-additional-vsphere_installation-config-parameters-vsphere)

### CODE REVIEW

File: modules/installation-configuration-parameters.adoc
Comment (skopacz1) [inline modules/installation-configuration-parameters.adoc:1]: I used the following logic to selectively include existing vSphere config parameters in the Agent docs:

- The entire vSphere section, which previously was included only in the vSphere configuration page, has been modified to be included in both the vSphere and Agent configuration pages ([here](https://github.com/openshift/openshift-docs/pull/70551/files#diff-14ebe576efb84b2f58b32a0de2996fced237204ae9b091e3cafbd40b65c1ef31R2423) and [here](https://github.com/openshift/openshift-docs/pull/70551/files#diff-14ebe576efb84b2f58b32a0de2996fced237204ae9b091e3cafbd40b65c1ef31R2648)) 
- Within that section of vSphere content, I used vSphere-only conditionals to exclude parameters that only belong in the vSphere page and not the Agent page. ([Here's an example](https://github.com/openshift/openshift-docs/pull/70551/files#diff-14ebe576efb84b2f58b32a0de2996fced237204ae9b091e3cafbd40b65c1ef31R2484-R2498) of content that remains in the vSphere page but does not appear in the Agent page)

**Comment by rwsu** [comment]
Hi @skopacz1. I think the existing vsphere config parameters are missing details. Under platform.vsphere.vcenters, there are additional parameters like server, user, password, port, and datacenters. Under platform.vsphere.failureDomains.topology, there are additional parameters like computeCluster, datacenter, datastore, resourcePool, and folder that should be described in greater detail. 

Screenshot 2024-02-06 at 2 27 17 PM
Screenshot from this page: https://docs.openshift.com/container-platform/4.14/installing/installing_vsphere/installing-vsphere.html#installation-vsphere-config-yaml_installing-vsphere

File: modules/installation-configuration-parameters.adoc [hunk 1]
```diff
@@ -2689,6 +2759,35 @@ Additional VMware vSphere configuration parameters are described in the followin
       datacenters:
 |Lists and defines the datacenters where {product-title} virtual machines (VMs) operate. The list of datacenters must match the list of datacenters specified in the `failureDomains` field.
 |String
+
+|platform:
+  vsphere:
+    vcenters:
+      password:
```
Comment (rvanderp3) [inline modules/installation-configuration-parameters.adoc:2766]: each vCenter is a separate entry in the vcenters slice. this allows multiple vCenters to be defined potentially(though its not supported at the moment).
```suggestion
      - password: <pw>
        port: 443
        server: <vcenter fqdn>
```

File: modules/installation-configuration-parameters.adoc [hunk 2]
```diff
@@ -2643,6 +2658,59 @@ Additional VMware vSphere configuration parameters are described in the followin
...
+|platform:
+  vsphere:
+    failureDomains:
+      topology:
+        folder:
+|Optional. The absolute path of an existing folder where the installation program creates the virtual machines, for example, `/<datacenter_name>/vm/<folder_name>/<subfolder_name>`.
+If you do not provide this value, the installation program creates a top-level folder in the datacenter virtual machine folder that is named with the infrastructure ID.
+If you are providing the infrastructure for the cluster and you do not want to use the default `StorageClass` object, named `thin`, you can omit the `folder` parameter from the `install-config.yaml` file.
+|String
+
```
Comment (rvanderp3) [inline modules/installation-configuration-parameters.adoc:2713]: see https://docs.openshift.com/container-platform/4.14/installing/installing_vsphere/installing-vsphere-installer-provisioned-customizations.html#configuring-vsphere-regions-zones_installing-vsphere-installer-provisioned-customizations for the schema to use for these fields.

File: modules/installation-configuration-parameters.adoc (hunk 1 above)
Comment (skopacz1) [reply modules/installation-configuration-parameters.adoc:2766]: That makes total sense. Each param is depicted this way because the install writers changed the old config format in the reference tables (e.g. `platform.vsphere.vcenters.password`) to this newer one that saves horizontal space. We're relying on the description/values column to convey things like what's optional and what needs to go together as a set. 

For example, for the compute section of the install-config, instead of documenting `architecture`, `name`, `platform`, etc like this, we rely on the value field of `compute.architecture` to describe the structure: "[Array of MachinePool objects](https://docs.openshift.com/container-platform/4.14/installing/installing_vsphere/installation-config-parameters-vsphere.html#installation-configuration-parameters-optional_installation-config-parameters-vsphere:~:text=the%20compute%20nodes.-,Array%20of%20MachinePool%20objects.,-compute%3A%0A%20%20architecture%3A)"

File: modules/installation-configuration-parameters.adoc [hunk 3]
```diff
@@ -2634,6 +2641,29 @@ Additional VMware vSphere configuration parameters are described in the followin
...
 |Establishes the relationships between a region and zone. You define a failure domain by using vCenter objects, such as a `datastore` object. A failure domain defines the vCenter location for {product-title} cluster nodes.
 |String
+endif::vsphere[]
+ifdef::agent[]
+
+|platform:
+  vsphere:
+    failureDomains:
+|Establishes the relationships between a region and zone. You define a failure domain by using vCenter objects, such as a `datastore` object. A failure domain defines the vCenter location for {product-title} cluster nodes.
+|An array of vCenter configuration objects.
```
Comment (rwsu) [inline modules/installation-configuration-parameters.adoc:2651]: I think vCenter is not correct here.

```suggestion
|An array of failure domain configuration objects.
```

**Review by rwsu — CHANGES_REQUESTED** [review]

**Review by rwsu — COMMENTED** [review]
/lgtm

**Comment by skopacz1** [comment]
@mhanss could you PTAL when you have a chance?

File: modules/installation-configuration-parameters.adoc [hunk 4]
```diff
@@ -2634,6 +2641,29 @@ Additional VMware vSphere configuration parameters are described in the followin
...
+  vsphere:
+    failureDomains:
+|Establishes the relationships between a region and zone. You define a failure domain by using vCenter objects, such as a `datastore` object. A failure domain defines the vCenter location for {product-title} cluster nodes.
+|An array of failure domain configuration objects.
+
+|platform:
+  vsphere:
+    failureDomains:
+      name:
+|The name of the vCenter server.
```
Comment (mhanss) [inline modules/installation-configuration-parameters.adoc:2657]: I think it should be 'failure domain name' instead of 'vCenter server'.

File: modules/installation-configuration-parameters.adoc [hunk 5]
```diff
@@ -2642,6 +2672,61 @@ Additional VMware vSphere configuration parameters are described in the followin
...
+====
+|String
+
+|platform:
+  vsphere:
+    failureDomains:
+      topology:
+        resourcePool:
+|Optional. The absolute path of an existing resource pool where the installation program creates the virtual machines, for example, `/<datacenter_name>/host/<cluster_name>/Resources/<resource_pool_name>/<optional_nested_resource_pool_name>`.
+If you do not specify a value, resources are installed in the root of the cluster, for example `/example_datacenter/host/example_cluster/Resources`.
```
Comment (mhanss) [inline modules/installation-configuration-parameters.adoc:2717]: For agent-based installations, users must specify the resource pool where the VMs are created because the installer does not create the VMs in this scenario.

File: modules/installation-configuration-parameters.adoc [hunk 6]
```diff
@@ -2642,6 +2672,61 @@ Additional VMware vSphere configuration parameters are described in the followin
...
+|String
+
+|platform:
+  vsphere:
+    failureDomains:
+      topology:
+        folder:
+|Optional. The absolute path of an existing folder where the installation program creates the virtual machines, for example, `/<datacenter_name>/vm/<folder_name>/<subfolder_name>`.
+If you do not provide this value, the installation program creates a top-level folder in the datacenter virtual machine folder that is named with the infrastructure ID.
+If you are providing the infrastructure for the cluster and you do not want to use the default `StorageClass` object, named `thin`, you can omit the `folder` parameter from the `install-config.yaml` file.
```
Comment (mhanss) [inline modules/installation-configuration-parameters.adoc:2727]: For agent-based installations, users must specify the folder where the VMs are created because the installer does not create the VMs in this scenario.

File: modules/installation-configuration-parameters.adoc (hunk 5 above)
Comment (skopacz1) [reply modules/installation-configuration-parameters.adoc:2717]: If I'm interpreting this correctly, I'll get rid of the last line about "not specifying a value". I'm assuming it's still technically an Optional parameter for ABI, but let me know if that's not the case.

File: modules/installation-configuration-parameters.adoc (hunk 6 above)
Comment (skopacz1) [reply modules/installation-configuration-parameters.adoc:2727]: If I'm interpreting this right, I'll get rid of line 2726 about "not providing the value". I'm assuming the line below it can stay and that this parameter is still Optional for ABI, but let me know if that's not the case.

File: modules/installation-configuration-parameters.adoc [hunk 7]
```diff
@@ -2642,6 +2672,63 @@ Additional VMware vSphere configuration parameters are described in the followin
...
+For more information, see "VMware vSphere region and zone enablement".
+====
+|String
+
+|platform:
+  vsphere:
+    failureDomains:
+      topology:
+        resourcePool:
+|Optional. The absolute path of an existing resource pool where the installation program creates the virtual machines, for example, `/<datacenter_name>/host/<cluster_name>/Resources/<resource_pool_name>/<optional_nested_resource_pool_name>`.
```
Comment (mhanss) [inline modules/installation-configuration-parameters.adoc:2716]: In case of ABI, the user creates the machines, not the installer program.
```suggestion
|Optional. The absolute path of an existing resource pool where the user has created the virtual machines, for example, `/<datacenter_name>/host/<cluster_name>/Resources/<resource_pool_name>/<optional_nested_resource_pool_name>`.
```

File: modules/installation-configuration-parameters.adoc [hunk 8]
```diff
@@ -2642,6 +2672,63 @@ Additional VMware vSphere configuration parameters are described in the followin
...
+
+|platform:
+  vsphere:
+    failureDomains:
+      topology:
+        folder:
+|Optional. The absolute path of an existing folder where the installation program creates the virtual machines, for example, `/<datacenter_name>/vm/<folder_name>/<subfolder_name>`.
+// Commenting out the line below because it doesn't apply to Agent, but it may be needed when this content is brought into the regular vSphere docs.
+// If you do not provide this value, the installation program creates a top-level folder in the datacenter virtual machine folder that is named with the infrastructure ID.
+If you are providing the infrastructure for the cluster and you do not want to use the default `StorageClass` object, named `thin`, you can omit the `folder` parameter from the `install-config.yaml` file.
```
Comment (mhanss) [inline modules/installation-configuration-parameters.adoc:2729]: In case of ABI, the user creates the machines, not the installer program. Also, I am unable to create an ISO without providing the folder, so it is a required field. Since it is a required field for ABI, we also need to modify this line "you can omit the `folder` parameter from the `install-config.yaml` file."
cc: @rwsu

Comment (skopacz1) [reply modules/installation-configuration-parameters.adoc:2729]: Sounds good, I modified the first line to say "folder where the user has created the virtual machines". 

Since the last sentence is based on omitting a required parameter, which we can't do, should I get rid of the sentence entirely, or is there a different way to perform the functionality described there?

Comment (skopacz1) [reply modules/installation-configuration-parameters.adoc:2729]: Per conversation on Slack, I'll remove that last sentence entirely and double check with Richard once it's removed.

Comment (rwsu) [reply modules/installation-configuration-parameters.adoc]: Manoj is correct. folder is not optional and is a required parameter. I think line 2729 should be commented out.

File: modules/installation-configuration-parameters.adoc [hunk 9]
```diff
@@ -2682,13 +2772,53 @@ Additional VMware vSphere configuration parameters are described in the followin
     vcenters:
 |Lists any fully-qualified hostname or IP address of a vCenter server.
 |String
+endif::vsphere[]
+ifdef::agent[]
+
+|platform:
+  vsphere:
+    vcenters:
+|Configures the connection details for services to communicate with vCenter.
```
[the remark this answers is not in the document]
Comment (michaelryanpeter) [reply modules/installation-configuration-parameters.adoc:2781]: Maybe "Configures the connection details so that services can communicate with vCenter"?

**Review by michaelryanpeter — COMMENTED** [review]
I left a couple of things to consider. Please reach out if you want to discuss anything. Great work!

/label peer-review-done
/remove-label peer-review-needed
/remove-label peer-review-in-progress

File: modules/installation-configuration-parameters.adoc (hunk 9 above)
[the remark this answers is not in the document]
Comment (skopacz1) [reply modules/installation-configuration-parameters.adoc:2781]: I think either of your suggestions work, although I'm not familiar enough with the content to know if the first one is still totally accurate. I think I prefer your second suggestion and will go with that one.

File: modules/installation-configuration-parameters.adoc [hunk 10]
```diff
@@ -2603,17 +2603,24 @@ Valid names include:
 --
 endif::ibm-cloud[]
 
-ifdef::vsphere[]
-
+ifdef::agent,vsphere[]
```
Comment (JoeAldinger) [inline modules/installation-configuration-parameters.adoc:2606]: I'm not sure I understand why you opened this ifdef here and then also on line 2616. I think my question is do you need line 2616? Just checking/learning: is it because you close this opening one down on line 2958?

**Review by JoeAldinger — COMMENTED** [review]
One quick question just to make sure things are good.

File: modules/installation-configuration-parameters.adoc (hunk 10 above)
Comment (skopacz1) [reply modules/installation-configuration-parameters.adoc:2606]: So technically this ifdef and the one on line 2616 are different because they check for slightly different conditions. This ifdef, which is closed on line 2958, basically says "Rule #1: all of the content within these lines can be included if `agent OR vsphere` is true, unless there's an exception".

The ifdef on line 2616 (which closes on line 2622) is basically treated like an exception to that `agent OR vsphere` condition, because it `ONLY` includes content for agent. It's basically saying "here's an exception to rule #1, the content from lines 2616 to 2622 can be included `ONLY` if `agent` is true, but can't be included in vsphere". Let me know if that makes sense, or if you'd like me to explain further.

TLDR the ifdef on line 2606 makes a general rule and the ifdefs like the ones on lines 2616 and 2623 make exceptions to that rule in order to be more specific

Comment (JoeAldinger) [reply modules/installation-configuration-parameters.adoc:2606]: I'm picking it up now. There just may be hope for me yet 😄 . Merging
